### PR TITLE
bunch of normal bugfix and some little improvements

### PR DESCRIPTION
**Pull request by pandaro**

### CODE REVIEW

File: buildsitehandler.lua [hunk 1]
```diff
@@ -358,6 +358,19 @@ function BuildSiteHandler:DontBuildOnMetalOrGeoSpots()
...
+		p = ai.buildsitehandler:ClosestBuildSpot(builder, nanoPos, utype)
+		if p then 
+			EchoDebug('found Position for near nano turret at: ' .. nanoPos.x ..' ' ..nanoPos.z)
+			break 
+		end
+
+	end
+	return p
+end
+
```
Comment (pandaro) [inline buildsitehandler.lua:373]: search for all the nano maybe can get computation problems... so maybe is better to save the position in order of building and start from the end of the list, in this manner we exclude the nanoturrets blocked in in a forest of other nano.
An  extreme  solution, if we noticed a great computing work, is save and use only the last one nano turret builded.
Try with this config please, or say me what do you think

File: buildsitehandler.lua [hunk 2]
```diff
@@ -374,7 +374,7 @@ end
 function BuildSiteHandler:BuilNearLastNano(builder,utype)
 	local p = nil
 	if self.ai.lastNanoBuild then
-		p = ai.buildsitehandler:ClosestBuildSpot(builder, nanoPos, utype)
+		p = ai.buildsitehandler:ClosestBuildSpot(builder, self.ai.lastNanoBuild, utype)
 	end
 	return p
 end
```
Comment (pandaro) [inline buildsitehandler.lua:380]: passed to last nano

File: factorybuildershandler.lua [hunk 3]
```diff
@@ -60,25 +60,35 @@ function FactoryBuildersHandler:PrePositionFilter()
...
 			buildMe = false 
 		end
 		if not buildMe and mtype == 'air' and ai.haveAdvFactory and ai.factoryBuilded['air'][1] > 0 and ai.factoryBuilded['air'][1] < 3 and isAdvanced then
-			self:EchoDebug('force build t2 air if you have t1 air and a t2 of another type')
+			self:EchoDebug(factoryName .. ' force build t2 air if you have t1 air and a t2 of another type')
 			buildMe = true
 		end
+		if buildMe and self.ai.factoriesAtLevel[1] and mtype == 'air' and isAdvanced and not ai.haveAdvFactory then
+			for index, factory in pairs(self.ai.factoriesAtLevel[1]) do
+				if unitTable[factory.unit:Internal():Name()].mtype ~= 'air' then
```
Comment (eronoobos) [inline factorybuildershandler.lua:84]: @pandaro the mtype of a factory isn't what mtype it produces. should be
`if factoryMobilities[factory.unit:Internal():Name()][1] ~= 'air' then`

Comment (eronoobos) [inline factorybuildershandler.lua:84]: https://github.com/eronoobos/BABAR-The-Shardifant/commit/e4fdd6a825e6509ca2957205700b56a7e027480f

**Pull request merged**